### PR TITLE
[Multisig] Rework of SigIndices

### PR DESCRIPTION
**Pull request by c4t-ag**
## Why this should be merged
The current solution has some hacks to come around the issue with sigIdy and addressing msig alias owned funds
This PR removes the hacks, and introduces separate from and signers in API methods

## How this works
There are 2 major changes to get things simplified

1. Allow different sets for From (spendable UTXOs) ans Signers (could be different to From for both multisig and avax ControlGroup) This let us specify things like: Spend funds from msig alias M and sign with addresses [A].

2. CaminoCredential contains next to the array of signatures a set of sigIdxs which are used to overwrite the sigIdxs in input.
This allows us to specify new inputIndices without chanaging the TX hash.
The len of this new sigIdxs array can be different compared to the len of input sigIdxs, reason is that there could be different ways how to verify owners with nested aliases. 
This change makes it also possible to remove the math.MaxUINT32 hack for signature indices

## How this was tested
Existing and new Unit tests

### CODE REVIEW

File: vms/platformvm/camino_service.go [hunk 1]
```diff
@@ -74,6 +75,8 @@ func (response GetBalanceResponseWrapper) MarshalJSON() ([]byte, error) {
 
 // GetBalance gets the balance of an address
 func (s *CaminoService) GetBalance(_ *http.Request, args *GetBalanceRequest, response *GetBalanceResponseWrapper) error {
+	s.vm.ctx.Log.Debug("Platform: GetBalance called")
```
Comment (pnowosie) [inline vms/platformvm/camino_service.go:78]: Is this left over from debugging or it's needed?

File: vms/platformvm/camino_service.go [hunk 2]
```diff
@@ -408,6 +416,13 @@ func (s *CaminoService) Spend(_ *http.Request, args *SpendArgs, response *SpendR
 		}
 	}
 
+	bytes, err = txs.Codec.Marshal(txs.Version, owners)
+	if err != nil {
+		return fmt.Errorf("%w: %s", errSerializeOwners, err)
+	}
+	if response.Owners, err = formatting.Encode(args.Encoding, bytes); err != nil {
+		return fmt.Errorf("%w: %s", errSerializeOwners, err)
```
Comment (pnowosie) [inline vms/platformvm/camino_service.go:424]: Both are errors so should be `%w` (also please fix prev occurences)
```suggestion
		return fmt.Errorf("%w: %w", errSerializeOwners, err)
```

File: vms/platformvm/camino_service.go [hunk 3]
```diff
@@ -703,48 +680,93 @@ func (s *Service) GetLastAcceptedBlock(r *http.Request, _ *struct{}, reply *api.
...
 
 	if err = user.Close(); err != nil {
 		return nil, err
 	}
-	return privKeys, nil
+	return keys, nil
 }
 
-func (s *Service) getFakeKeys(args *api.JSONFromAddrs) (*secp256k1fx.Keychain, error) {
+func (s *Service) getFakeKeys(from *api.JSONFromAddrs) ([]*crypto.PrivateKeySECP256K1R, error) {
```
Comment (pnowosie) [inline vms/platformvm/camino_service.go:747]: I'm missing the purpose why are we faking the keys.
Please consider to add comment to this fn.

File: vms/secp256k1fx/camino_fx.go [hunk 4]
```diff
@@ -6,23 +6,23 @@ package secp256k1fx
...
-	errNotSecp256Cred   = errors.New("expected secp256k1 credentials")
-	errWrongOutputType  = errors.New("wrong output type")
-	errMsigCombination  = errors.New("msig combinations not supported")
-	errNotAliasGetter   = errors.New("state isn't msig alias getter")
-	errTooFewSigIndices = errors.New("too few signature indices")
+	errNotSecp256Cred  = errors.New("expected secp256k1 credentials")
+	errWrongOutputType = errors.New("wrong output type")
+	errMsigCombination = errors.New("msig combinations not supported")
+	errNotAliasGetter  = errors.New("state isn't msig alias getter")
+	errWrongSigIdxsLen = errors.New("multisig sigidx count invalid")
```
Comment (pnowosie) [inline vms/secp256k1fx/camino_fx.go]:
```suggestion
	errWrongSigIdxsLen = errors.New("multisig sigidx count mismatch")
```

File: vms/secp256k1fx/camino_fx.go [hunk 5]
```diff
@@ -268,3 +260,39 @@ func (fx *Fx) verifyMultisigUnorderedCredentials(tx UnsignedTx, creds []verify.V
 
 	return nil
 }
+
+// SplitPrivateKeys splits an array of PrivateKeys into `from` and `signers`
+// The delimiter is a `nil` PrivateKey.
+// If no delimiter exists, the given PrivateKeys are used for both from and signing
+// Having different sets of `from` and `signer` allow multisignatuew alias feature
```
Comment (pnowosie) [inline vms/secp256k1fx/camino_fx.go]:
```suggestion
// Having different sets of `from` and `signer` allows multisignature feature
```

File: vms/secp256k1fx/camino_fx.go [hunk 6]
```diff
@@ -268,3 +257,39 @@ func (fx *Fx) verifyMultisigUnorderedCredentials(tx UnsignedTx, creds []verify.V
...
+	// find nil key which splits froms and signer
+	splitIndex := len(keys)
+	for index, key := range keys {
+		if key == nil {
+			splitIndex = index
+			break
+		}
+	}
+
+	if splitIndex == len(keys) {
```
Comment (pnowosie) [inline vms/secp256k1fx/camino_fx.go:274]: We can remove this entirerly and make just 2 loops as below but both indexes on `keys` slice:

```go
	for index := 0; index < splitIndex; index++ {
		from.Add(keys[index].PublicKey().Address())
	}
	
	//...
	for index := splitIndex+1; index < len(keys); index++ {
		signer = append(signer, keys[index])
	}
```

**Review by pnowosie — CHANGES_REQUESTED** [review]

File: vms/platformvm/utxo/camino_locked.go [hunk 7]
```diff
@@ -210,28 +213,27 @@ func (h *handler) Lock(
...
 	default:
-		return nil, nil, nil, errInvalidTargetLockState
+		return nil, nil, nil, nil, errInvalidTargetLockState
 	}
 
-	addrs := set.NewSet[ids.ShortID](len(keys)) // The addresses controlled by [keys]
-	for _, key := range keys {
-		addrs.Add(key.PublicKey().Address())
-	}
+	addrs, signer := secp256k1fx.SplitPrivateKeys(keys)
```
Comment (evlekht) [inline vms/platformvm/utxo/camino_locked.go]:
```suggestion
	spendFromAddrs, signer := secp256k1fx.SplitPrivateKeys(keys)
```
not sure, but imo would give better explanation about splitting

File: vms/platformvm/camino_service.go (hunk 1 above)
Comment (c4t-ag) [reply vms/platformvm/camino_service.go:78]: It is in every (avax and our) api call, if we arrive strange traffic in node we can see in debug log if something is called very often. So no, no debugging, it was simply missing....

File: vms/secp256k1fx/camino_credential.go [hunk 8]
```diff
@@ -4,34 +4,37 @@
...
+	// Overwrites sigIdx in inputs of a tx
+	// Must be same length as sigIdx in input
+	SigIdxs []uint32 `serialize:"true"`
 }
 
 func (cr *Credential) Signatures() [][crypto.SECP256K1RSigLen]byte {
 	return cr.Sigs
 }
 
+func (*Credential) SignatureIndices() []uint32 {
```
Comment (evlekht) [inline vms/secp256k1fx/camino_credential.go:36]: Is it ok, that MultisigCredential will use the same SignatureIndices() method with nil in return?

File: vms/secp256k1fx/camino_fx.go [hunk 9]
```diff
@@ -268,3 +257,39 @@ func (fx *Fx) verifyMultisigUnorderedCredentials(tx UnsignedTx, creds []verify.V
 
 	return nil
 }
+
+// SplitPrivateKeys splits an array of PrivateKeys into `from` and `signers`
+// The delimiter is a `nil` PrivateKey.
+// If no delimiter exists, the given PrivateKeys are used for both from and signing
+// Having different sets of `from` and `signer` allow multisignatuew alias feature
+func SplitPrivateKeys(keys []*crypto.PrivateKeySECP256K1R) (set.Set[ids.ShortID], []*crypto.PrivateKeySECP256K1R) {
```
Comment (evlekht) [inline vms/secp256k1fx/camino_fx.go]: Maybe renamed SplitPrivateKeys to something like SplitIntoSiginersAndFromAddrs ?
Cause split sounds very general - like split thing into many things
But this funcs does more specific task (and comments describe this specific)

In general, this concept of key array where part of keys are just as addrs feels to shady, but I guess thats done in order to do less avax changes

File: vms/secp256k1fx/camino_fx.go (hunk 4 above)
Comment (c4t-ag) [reply vms/secp256k1fx/camino_fx.go]: err deleted (not longer used)

File: vms/secp256k1fx/camino_fx.go (hunk 6 above)
Comment (c4t-ag) [reply vms/secp256k1fx/camino_fx.go:274]: But only with an inner branch (if / else), and I don't like it -> stays like it is

File: vms/secp256k1fx/camino_keychain.go [hunk 10]
```diff
@@ -20,23 +20,15 @@ var (
...
-const (
-	TFVerify TFResult = iota
-	TFSkip
-	TFContinue
-	TFError
-)
-
 // Spend attempts to create an input from outputowners which can contain multisig aliases
-func (kc *Keychain) SpendMultiSig(out verify.Verifiable, time uint64, msigIntf interface{}) (
+// If collect is true, all addresses prefixed by a header with num and threashold is returned
```
Comment (evlekht) [inline vms/secp256k1fx/camino_keychain.go]: What is collect?

File: vms/platformvm/camino_service.go (hunk 2 above)
Comment (c4t-ag) [reply vms/platformvm/camino_service.go:424]: Double %w not supported by golang

File: vms/secp256k1fx/camino_keychain.go [hunk 11]
```diff
@@ -143,47 +134,32 @@ func TraverseOwners(out *OutputOwners, msig AliasGetter, callback TraverserOwner
...
+					)
+					if err != nil {
+						return 0, err
+					}
+					if success {
+						currentStack.verified++
+						addrVerified++
+
+						if addrVerified > MaxSignatures {
+							return 0, errTooManySignatures
```
Comment (evlekht) [inline vms/secp256k1fx/camino_keychain.go:159]: isn't error the opposite?

File: vms/secp256k1fx/camino_fx.go (hunk 9 above)
Comment (c4t-ag) [reply vms/secp256k1fx/camino_fx.go]: Exactly I started some time ago with an additional parameter but it lead into hell on earth

File: vms/secp256k1fx/camino_keychain.go (hunk 11 above)
Comment (c4t-ag) [reply vms/secp256k1fx/camino_keychain.go:159]: Looks good for me, there is simply for now a Limit of verifiable addresses (MaxSignatures constant)

Comment (evlekht) [reply vms/secp256k1fx/camino_keychain.go:159]: ah, got it

File: vms/secp256k1fx/camino_credential.go (hunk 8 above)
Comment (c4t-ag) [reply vms/secp256k1fx/camino_credential.go:36]: It's interface implementation, so yes, that should be fine, oe am I wrong?
Normal avax Credential doesn't have sigIdx overwrite.....

Comment (evlekht) [reply vms/secp256k1fx/camino_credential.go:36]: I mean, that `MultisigCredential.SignatureIndices()` will always return nil - is this intended?

File: vms/platformvm/camino_service.go (hunk 2 above)
Comment (pnowosie) [reply vms/platformvm/camino_service.go:424]: Ouch! Thx

File: vms/secp256k1fx/camino_fx.go (hunk 6 above)
Comment (pnowosie) [reply vms/secp256k1fx/camino_fx.go:274]: Oh, I see, the return values diff. 👍

**Review by pnowosie — DISMISSED** [review]

File: vms/secp256k1fx/camino_credential.go (hunk 8 above)
Comment (c4t-ag) [reply vms/secp256k1fx/camino_credential.go:36]: Thanks man, yes, this is for sure NOT intended :-)

**Review by evlekht — APPROVED** [review]

**Pull request merged**